operators/src/pro/ml/xgboost.rs

```
@@ -0,0 +1,694 @@
+
+        Ok(initialized_operator.boxed())
+    }
+
+    span_fn!(XgboostOperator);
+}
+
+impl InitializedRasterOperator for InitializedXgboostOperator {
+    fn result_descriptor(&self) -> &RasterResultDescriptor {
+        &self.result_descriptor
+    }
+
+    fn query_processor(&self) -> Result<TypedRasterQueryProcessor> {
+        let vec_of_rqps: Vec<Box<dyn RasterQueryProcessor<RasterType = f32>>> = self
+            .sources
+            .iter()
+            .map(|init_raster| {
+                let typed_raster_qp = match init_raster.query_processor() {
+                    Ok(r) => r,
+                    Err(e) => {
+                        return Err(Error::XGBoost {
+                            source: XGBoostModuleError::OperatorsError {
+                                source: Box::new(e),
+                            },
+                        })
+                    }
+                };
+                let converted = typed_raster_qp.into_f32();
+                Ok(converted)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(QueryProcessorOut(Box::new(XgboostProcessor::new(
+            vec_of_rqps,
+            self.model_sub_path.clone(),
+            self.no_data_value,
+        ))))
+    }
+}
+
+struct XgboostProcessor<Q, P>
+where
+    Q: RasterQueryProcessor<RasterType = P>,
+{
+    sources: Vec<Q>,
+    model_file: PathBuf,
+    no_data_value: f32,
+}
+
+impl<Q, P> XgboostProcessor<Q, P>
+where
+    Q: RasterQueryProcessor<RasterType = P>,
+    P: Pixel,
+{
+    pub fn new(sources: Vec<Q>, model_file_path: PathBuf, no_data_value: f32) -> Self {
+        Self {
+            sources,
+            model_file: model_file_path,
+            no_data_value,
+        }
+    }
+
+    async fn process_tile_async(
+        &self,
+        bands_of_tile: Result<Vec<RasterTile2D<P>>>,
+        pool: Arc<ThreadPool>,
+    ) -> Result<BaseTile<GridOrEmpty<GridShape<[usize; 2]>, f32>>> {
+        let bot = match bands_of_tile {
```

- maybe you should do the error handling before calling this method. Then you can pass a `Vec<RasterTile2D<P>>` instead of a Result
- why is the error called `NoInputData`? Shouldn't we propagate the error that occured in the source that produced the error tile?
- collecting the input tiles into a vector should better also be done in the spawned task and not in the async code

---

https://github.com/geo-engine/geoengine/pull/639/commits/0c5afdf29e30de1f9614430b19434949b8434f3e

I struggled a lot with finding a way to move the error handling out of the function and into the future closure. I'm sure there is a better way for this?

---

Ok so I guess this part of the code can be summarized as 'learning curve'. I hope the changes I made now are making things a little bit more concise https://github.com/geo-engine/geoengine/pull/639/commits/1e11b8ab7d01829e7190ffd2475331f073d8e13c. 
I was taking a bit of inspiration by this part of the time shift operator here:
https://github.com/geo-engine/geoengine/blob/826367d205677a7d1e30cba493910bb10b0304c6/operators/src/processing/time_shift.rs#L412-L431

---

do you mean `Ok(processed_tile.await?)`? That's fine, the closure within `.then` will only be called once a new item is ready. So the `await` happens in the resulting stream and not while creating the stream.